Source/Core/InputCommon/GCAdapter.cpp

```diff
@@ -203,14 +203,26 @@ static void ReadThreadFunc()
     std::array<u8, CONTROLER_INPUT_PAYLOAD_EXPECTED_SIZE> input_buffer;
 
     int payload_size = 0;
-    const int error =
-        libusb_interrupt_transfer(s_handle, s_endpoint_in, input_buffer.data(),
-                                  int(input_buffer.size()), &payload_size, USB_TIMEOUT_MS);
+    int error = libusb_interrupt_transfer(s_handle, s_endpoint_in, input_buffer.data(),
+                                          int(input_buffer.size()), &payload_size, USB_TIMEOUT_MS);
     if (error != LIBUSB_SUCCESS)
     {
       ERROR_LOG_FMT(CONTROLLERINTERFACE, "Read: libusb_interrupt_transfer failed: {}",
                     LibusbUtils::ErrorWrap(error));
     }
+    if (error == LIBUSB_ERROR_IO)
+    {
+      // s_read_adapter_thread_running is cleared by the joiner, not the stopper.
+
+      // Reset the device, which may trigger a replug.
+      error = libusb_reset_device(s_handle);
+      ERROR_LOG_FMT(CONTROLLERINTERFACE, "Read: libusb_reset_device: {}",
+                    LibusbUtils::ErrorWrap(error));
+      if (error != 0)
+      {
+        break;
```

Comment:
What would the user have to do to re-enable the GC Adapter if this `break;` is hit? I get how you arrived at this but it's kinda weird how the thread just dies here without informing anyone about this failure.

Reply:
Oof. Normally `Reset()` checks `if (s_read_adapter_thread_running.TestAndClear()) s_read_adapter_thread.join();`, which happens when the USB device is unplugged.

If I ignore `libusb_reset_device()` return value and always assume it fails, I found that switching away from GC Adapter and back doesn't work (and the controller dialog wrongly states "Adapter Detected"). I think there needs to be some way for the GC thread to communicate to the GUI whether it's stopped on its own or not. Unfortunately [`std::thread`](https://en.cppreference.com/w/cpp/thread/thread) lacks a non-blocking "has terminated" API (and only offers blocking join).

One possible workaround is to add a new atomic boolean variable for "is thread running". I tried setting `s_adapter_error = static_cast<libusb_error>(error); s_status = AdapterStatus::Error` before breaking out of the loop, but this shows a message on the UI "Error Opening Adapter: Input/Output Error" (suggesting an error on *opening* the adapter, not recovering from sleep-wake), and switching away from the adapter and back still shows the same error. Unfortunately even unplugging and plugging the adapter does not help for some reason.

I'll have to trace what's going on with the threads further, perhaps I didn't close a device somewhere. (Though does it really matter given that I've only ever seen this case by pretending that `libusb_reset_device()` fails?)